NDI6 SDK + DistroAV obs-studio plugin

globule655 opened the pull request:
<!--
New package ndi-6 was added. An existing version 4 already exists but as those version are not inter-compatible I thought more suitable to have different packages.
DistroAV has been added as an obs-studio plugin. It provides the ability to add NDI sources inside OBS Studio. Requires NDI runtime version 5 and above to work.
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->
ndi-6: https://ndi.video/tech/ndi6/
DistroAV: https://github.com/DistroAV/DistroAV

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [x] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

globule655 committed:
maintainers: add globule655

globule655 committed:
obs-studio-plugins.distroav: init at 6.0.0

This Version has major & large code changes.Starting v6 the plugin has
been re-branded DistroAV from OBS-NDI.
https://github.com/DistroAV/DistroAV/releases/tag/6.0.0

"obs-studio-plugins.distroav: correct file formatting"

"obs-studio-plugins.distroav: fixed fmt issue by adding tag"

obs-studio-plugins.distroav: run nixfmt on the file to fix formatting

osb-studio-plugins.distroav: changed lib reference to ndi-6

obs-studio-plugins.distroav: fix ndi-6 reference

globule655 committed:
ndi-6: init at 6.1.1

Major update from existing 4.x package.
Compatible with DistroAV obs-studio plugin
https://docs.ndi.video/all/getting-started/release-notes#ndi-6.1.1

SigmaSquadron commented:
Welcome to Nixpkgs! The official formatter for Nix code in Nixpkgs is `nixfmt`. You can format your files by running `nix-shell` in the Nixpkgs root directory and running `nixfmt /path/to/your/changed/files`

SigmaSquadron commented on pkgs/by-name/nd/ndi-6/package.nix:
```diff
@@ -0,0 +1,89 @@
+      "aarch64-linux"
+      "armv7l-linux"
+    ];
+    hydraPlatforms = [ ];
```
```suggestion
```
Hydra will never build this anyway, since it's `unfree`.

SigmaSquadron commented on pkgs/top-level/all-packages.nix:
```diff
@@ -18706,4 +18706,5 @@ with pkgs;
   wings = callPackage ../by-name/wi/wings/package.nix {
     erlang = erlang_25;
   };
+
```
```suggestion
```
Please delete this newline.

SigmaSquadron reviewed (commented):
Looks good overall! Don't forget that the commits need to be squashed per the [Commit Conventions](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#commit-conventions).

globule655 commented on pkgs/applications/video/obs-studio/plugins/distroav/default.nix:
```diff
@@ -0,0 +1,61 @@
+    sha256 = "sha256-pr/5XCLo5fzergIQrYFC9o9K+KuP4leDk5/oRe5ct9Q=";
+  };
+
+  patches = [ ./hardcode-ndi-path.patch ];
```
Good idea. It took me a while to figure it out myself :+1:

globule655 commented on pkgs/applications/video/obs-studio/plugins/distroav/default.nix:
```diff
@@ -0,0 +1,61 @@
+
+  postPatch = ''
+    # Add path (variable added in hardcode-ndi-path.patch
+    sed -i -e s,@NDI@,${ndi-6},g src/plugin-main.cpp
```
Did not know about this. Very good indeed

globule655 commented on pkgs/applications/video/obs-studio/plugins/distroav/default.nix:
```diff
@@ -0,0 +1,61 @@
+    license = licenses.gpl2;
+    maintainers = with maintainers; [ globule655 ];
+    platforms = platforms.linux;
+    hydraPlatforms = ndi-6.meta.hydraPlatforms;
```
OK thanks for the explanation. It was there on the template I used and didn't really know what it was about so I left it there. Removed now

globule655 commented on pkgs/by-name/nd/ndi-6/package.nix:
```diff
@@ -0,0 +1,89 @@
+      if [ -L "$i" ]; then continue; fi
+      patchelf --set-interpreter "$(cat $NIX_CC/nix-support/dynamic-linker)" "$i"
+    done
+    patchelf --set-rpath "${avahi}/lib:${stdenv.cc.libc}/lib" $out/bin/ndi-record
+    patchelf --set-rpath "${avahi}/lib:${stdenv.cc.libc}/lib" $out/bin/ndi-free-audio
```
Never used this so I will have to  look into it. I will leave this as is for now but will certainly read the docs for future iterations and/or contributions. Thanks for the tip :+1:

globule655 commented:
Thanks a lot for the review and great suggestions !
I think I fixed all the things pointed out and squashed the commits.
Let me know if there is anything else that needs to be changed

SigmaSquadron commented:
Good work! Just one thing: the commits were not fully squashed. There should be three commits:

1. `maintainers: add globule655` which only adds your maintainer entry to `maintainer-list.nix`.
2. `obs-studio-plugins.distroav: init at 6.0.0` which only adds the `distroav` package.
3. `ndi-6: init at 6.1.1` which only adds the `ndi-6` package.

globule655 commented:
> Good work! Just one thing: the commits were not fully squashed (you need to force push your branch after squashing). There should be three commits:
> 
> 1. `maintainers: add globule655` which only adds your maintainer entry to `maintainer-list.nix`.
> 2. `obs-studio-plugins.distroav: init at 6.0.0` which only adds the `distroav` package.
> 3. `ndi-6: init at 6.1.1` which only adds the `ndi-6` package.
> 
> All of the fixes should be squashed into the two latter commits.

Whoops my bad. Fixed now, only 3 commits.j
Thanks again

globule655 commented:
I just changed a little thing with the substitueInPlace command.
After thinking about it I thought it best to use --replace-fail as the plugin won't work if no changes are made by the command

globule655 commented:
Question to anybody in the know :
When checking the ofborg tests logs I see the following error
```Cannot nix-instantiate `obs-studio-plugins.distroav.passthru.tests` because:
error: attribute 'tests' in selection path 'obs-studio-plugins.distroav.passthru.tests' not found```

My understanding of the passthru.tests function is that it checks that the other package can be built with the current one as a dependency or something like that.
e.g : Can I build obs-studio-plugins.distroav which depends on ndi-6 ?

So here as it's failing I'm wondering 2 things :

* Did I understand incorrectly how I should write this ? I saw this test written exactly like that for other packages so I figured it was okay
* Is that test really necessary in the end ? I guess it's nice to have but can be omitted without much risk

Cheers for the new year's eve :beers:

globule655 commented:
Removed the passthru.tests directive for now as after reading the docs I didn't think it was something absolutely necessary and caused the pipelines check to hang

SigmaSquadron commented:
> Removed the passthru.tests directive for now as after reading the docs I didn't think it was something absolutely necessary and caused the pipelines check to hang

Don't worry about CI. OfBorg is dead, and it won't be running any tests anytime soon. Please re-add the tests if they weren't broken.

globule655 commented:
> > Removed the passthru.tests directive for now as after reading the docs I didn't think it was something absolutely necessary and caused the pipelines check to hang
> 
> Don't worry about CI. OfBorg is dead, and it won't be running any tests anytime soon. Please re-add the tests if they weren't broken.

Hey and thanks for the clarification !
I put back the test then as it was all fine, only the pipeline hanging made me doubt

Cheers

globule655 commented:
Hi there,
Anything else needed from me for this to be merged ?
What are the next steps ?

Thanks a lot

globule655 commented:
bump

globule655 commented:
Hi,
Is there a chance this will be merged some day ?
Something I still need to do before it happens ?

Thanks a lot

Eveeifyeve commented:
Consider making an alias to obs-ndi to distroav and eventually returning deprecating obs-ndi throw in 26.05.
